File: neetcode150/easy/20-Valid-Parentheses.py
#!/usr/bin/python3

class Solution:
  # Stack based solution
  def isValid(self, s: str) -> bool:

    stack = []

    # Iterate through all paren in given string
    for l in s:
      # If its an opening paren add it to the top of the stack
      if l in ['(', '[', '{']:
        stack.append(l)
      # If the stack is empty return false
      elif not stack:
        return False
      # Check the top of the stack for the matching paren and remove
      # it from the stack
      elif l == ')' and stack[-1] == '(':
        stack.pop()
      elif l == ']' and stack[-1] == '[':
        stack.pop()
      elif l == '{' and stack[-1] == '{':
        stack.pop()
      else:
        return False
    # if the stack is empty and all paren's are valid return true
    # necessary to catch edge cases like "[" or "(("
    if len(stack) == 0:
      return True
    return False

File: neetcode150/easy/20-Valid-Parentheses.cpp
#include <stack>
#include <string>

using namespace std;

/**
 * \c Solution
 * 
 * 1. Push open parens to stack
 * 2. On closed paren check if stack.top is the open paren
 * 
 * Time  : O(N)
 * Space : O(N)
 */
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;

        for (char p : s) {
            if (p == '(' || p == '[' || p == '{') {
                st.push(p);
            } else if (st.empty()) {
                return false;
            } else if (p == ')' && st.top() != '(') {
                return false;
            } else if (p == ']' && st.top() != '[') {
                return false;
            } else if (p == '}' && st.top() != '{') {
                return false;
            } else {
                st.pop();
            }
        }
        return st.empty();
    }
};